### PR TITLE
refactor: convert Validator to Class

### PR DESCRIPTION
**Pull request by alexlafroscia**
The "meat" of this change converts the existing Validator implementation into a class. Specifically, Validators are now Monoids that can be combined to create more complex validations:

> This refactors the concept of what a Validator actually is, changing it from a Function into a Class. This allows for co-locating all of the information about how to check a value and format the error message into a single location.
>
> The Validator class can be treated as a Monoid, meaning that it can be combined with other instances of itself to build up more complex Validators. Many operations that we supported previously can be thought of as combinations of other Validators, so I thought that formalizing that made sense. With base combinators like `and` and `or`, we can build up some of the type modifiers that we support like `arrayOf` or `oneOf`. All of these can be expressed as Validators, in terms of other Validators.

**Note:** This will conflict with some of the changes made in #88. I'll refactor whichever pull request lands later so that they're compatible.

### CODE REVIEW

**Comment by alexlafroscia** [comment]
I wanted to expose the base `Validator` class to allow users to extend it themselves and create custom validators, but I'm not sure that's possible with the current state of the `babel-plugin-filter-imports`. Those code trips up the plugin:

```javascript
import { Validator } from '@ember-decorators/argument';

class SomeValidator extends Validator {}
```

With good enough dead code elimination, it should be safe to support this; we filter out the use of `@argument` everywhere, which removes all usage of `SomeValidator`, which means that `Validator` is no longer used and can also be pruned. However, this doesn't seem to be supported by `babel-plugin-filter-imports` at this time and would be easy to implement in the future, should we figure out how to support it.

File: addon/-private/combinators/or.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+export class OrValidator extends BaseValidator {
+  constructor(...validators) {
+    super();
+
+    this.validators = validators;
+  }
+
+  check(value) {
+    return this.validators.reduce(
```
Comment (pzuraq) [inline addon/-private/combinators/or.js]: could probably do `this.validators.some(validator => validator.check(value))`

File: addon/-private/validators/any.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import BaseValidator from './-base';
+
+export default class AnyValue extends BaseValidator {
```
Comment (pzuraq) [inline addon/-private/validators/any.js]:
```suggestion
 export default class AnyValidator extends BaseValidator {
```

For consistency

**Review by pzuraq — APPROVED** [review]
👍 This looks good, I think the classes really clean things up in general, nice job!

File: addon/-private/combinators/or.js (hunk 1 above)
Comment (alexlafroscia) [reply addon/-private/combinators/or.js]: Good catch, thank you!

**Pull request merged**